lib/grant-types/authorization-code-grant-type.js

```diff
@@ -0,0 +1,170 @@
+  }
+
+  if (!client) {
+    throw new InvalidArgumentError('Missing parameter: `client`');
+  }
+
+  var scope = this.getScope(request);
+
+  return Promise.bind(this)
+    .then(function() {
+      return this.getAuthorizationCode(request, client);
+    })
+    .tap(function(code) {
+      return this.revokeAuthorizationCode(code);
+    })
+    .then(function(code) {
+      return this.saveToken(code.user, client, code.authorizationCode, scope);
+    });
+};
+
+/**
+ * Get the authorization code.
+ */
+
+AuthorizationCodeGrantType.prototype.getAuthorizationCode = function(request, client) {
+  if (!request.body.code) {
+    throw new InvalidRequestError('Missing parameter: `code`');
+  }
+
+  if (!is.vschar(request.body.code)) {
+    throw new InvalidRequestError('Invalid parameter: `code`');
+  }
+
+  return Promise.try(this.model.getAuthorizationCode, request.body.code)
+    .then(function(code) {
+      if (!code) {
+        throw new InvalidGrantError('Invalid grant: authorization code is invalid');
+      }
+
+      if (!code.client) {
+        throw new ServerError('Server error: `getAuthorizationCode()` did not return a `client` object');
+      }
+
+      if (!code.user) {
+        throw new ServerError('Server error: `getAuthorizationCode()` did not return a `user` object');
+      }
+
+      if (code.client.id !== client.id) {
+        throw new InvalidGrantError('Invalid grant: authorization code is invalid');
+      }
+
+      if (!(code.expiresAt instanceof Date)) {
+        throw new ServerError('Server error: `expiresAt` must be a Date instance');
+      }
+
+      if (code.expiresAt < new Date()) {
+        throw new InvalidGrantError('Invalid grant: authorization code has expired');
+      }
+
+      return code;
+    });
+};
+
+/**
+ * Revoke the authorization code.
+ *
+ * "The authorization code MUST expire shortly after it is issued to mitigate
+ * the risk of leaks. [...] If an authorization code is used more than once,
+ * the authorization server MUST deny the request."
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.2
+ */
+
+AuthorizationCodeGrantType.prototype.revokeAuthorizationCode = function(code) {
+  return Promise.try(this.model.revokeAuthorizationCode, code)
+    .then(function(code) {
+      if (!code) {
+        throw new InvalidGrantError('Invalid grant: authorization code is invalid');
+      }
+
+      if (!(code.expiresAt instanceof Date)) {
+        throw new ServerError('Server error: `expiresAt` must be a Date instance');
+      }
+
+      if (code.expiresAt >= new Date()) {
+        throw new ServerError('Server error: authorization code should be expired');
+      }
+
+      return code;
+    });
+};
+
+/**
+ * Save token.
+ */
+
+AuthorizationCodeGrantType.prototype.saveToken = function(user, client, authorizationCode, scope) {
+  return this.generateAccessToken()
+    .bind(this)
+    .then(function(accessToken) {
+      var token = {
```

Comment:
> This appears to be missing the refresh token and the user as well as the token expiration.

Indeed. I didn't have a need for it in my particular implementation and ended up not adding it. I will add it shortly.

> Also why is authorization code included as part of this?

The `authorizationCode` is sent to the `saveToken()` callback to allow users to associate a token with an authorization code. It is immensely helpful for server apps to determine which _accessToken_ was generated by which _authorizationCode_.